Add slideshow for Navier Stokes Module (#21591)
idaholab/moose: modules/navier_stokes/doc/intro_slides/slides/index.md

# The Navier-Stokes Module

!style halign=center fontsize=120%
!datetime today format=%B %Y

!---

# Overview

Main capabilities:

> - Incompressible fluid flows
> - Weakly-compressible fluid flows
> - Compressible fluid flows

Reviewer: Maybe merge together:
```suggestion
- Flow types: incompressible, weakly-compressible, and compressible
```

- Free flows and flows in porous media describing homogenized structures

> Utilizies the following techniques:

Reviewer 2: ```suggestion
Utilizes the following techniques:
```

> - Continuous Finite Element Discretization (maintained, but not intensely developed)

Reviewer: ```suggestion
- Stabilized Continuous Finite Element Discretization (maintained, but not intensely developed)
```

- Finite Volume Discretization (current development direction)

!---

> # The Navier-Stokes equations ([!citet](radman2021development))

Reviewer: Why do we cite Radman et al. as a reference for the NS equations? I thought their paper was two-phase. Please consider revising

Reviewer: The one that people generally cite is: K. Vafai, Handbook of Porous Media, CRC Press, Taylor & Francis, New York (2005)

Reviewer 3: ```suggestion
# The Porous Navier-Stokes equations ([!citet](radman2021development))
```

Author: I'll add the handbook too. I've read it but I think the formulation in Radman is the closest to what we have. It is general enough for single phase too. I'll keep them both.

  style=font-size:26pt

!style! fontsize=80%

- Conservation of +mass+:

!equation id=mass-eq
\frac{\partial \gamma \rho}{\partial t} + \nabla \cdot \left( \rho \vec{u} \right)  = 0

- Conservation of (linear) +momentum+:

!equation id=momentum-eq
\frac{\partial \rho  \vec{u}}{\partial t} + \nabla \cdot \left(\gamma^{-1}  \rho \vec{u} \otimes \vec{u}\right)
= \nabla \cdot \left(\gamma \mu_\text{eff} \left(\nabla\vec{u}_I + \left(\nabla \vec{u}_I\right)^T-\frac{2}{3}\nabla\cdot\vec{u}_I\mathbb{I}\right)\right) -\gamma\nabla p + \gamma  \rho \vec{g} -  W  \vec{u}_I

!style-end!

!row!
!col! width=30%

!style! fontsize=70%
- $\gamma$ - porosity
- $\vec{u}_I$ - real (or interstitial) velocity
- $\vec{u} \equiv \gamma \vec{u}_I$ - superficial velocity
!style-end!

!col-end!

!col! width=60%

!style! fontsize=70%
- $\rho = \rho(p,T)$ - density
- $\mu_{eff}$ - effective dynamic viscosity (laminar + turbulent)
- $p$ - pressure
- $W$ - friction tensor (using correlations)
!style-end!

!col-end!
!row-end!

!style! fontsize=80%
- If $\gamma=1$ then $\vec{u}=\vec{u}_I$ and we get the original Navier-Stokes equations back.
!style-end!

!---

> # The Navier-Stokes equations (contd.)

Reviewer 3: ```suggestion
# The Porous Navier-Stokes equations (contd.)
```

  style=font-size:26pt

- These are supplemented by the conservarion of +energy+:

  +Assumptions:+ Weakly-compressible fluid, kinteic energy of the fluid can be neglected

!equation id=enthalpy-eq

> \frac{\partial \gamma \rho c_p T}{\partial t} + \nabla \cdot \left(  \rho c_p T \vec{u}\right)
> = \nabla \cdot \left(\kappa_f \nabla T \right) - \alpha (T - T_s) + S_{ext}

Reviewer: ```suggestion
\frac{\partial \gamma \rho c_p T}{\partial t} + \nabla \cdot \left(  \rho c_p T \vec{u}\right)
= \nabla \cdot \left(\gamma \kappa_f \nabla T \right) - \gamma \alpha (T - T_s) + \gamma S_{ext}
```

Reviewer: The above is the mathematically correct formulation. Feel free to disregard if this is implemented differently in the code. I didn't check

Reviewer 2: for diffusivity it depends how you chose to factor in the effect of porosity. Both options are available in the kernel. 

For the S_ext term we dont have it in any simulation right now, it's conventions as well imo

For the alpha dT term we should discuss.

Author: I assumed gamma is in kappa_s, I followed the description from the M2 report. I'll leave it like that for now, we can change it later.

Author: Actually, I don't think we have gamma on alpha*dT in this formulation. I assumed it is some sort of alpha = alpha_hx * a_v where a_v is a vlumetric surface m2/m3.

Author: Also, I did not define what S_ext is, you can do whatever you want with it. I just added an example source here. In some cases the source term is already compensated with a volume fraction in other apps (I think this was the issue with Cole's input, no?).

!row!
!col! width=30%

!style! fontsize=80%
- $c_p$ - specific heat
- $T$ - fluid temperature
- $\kappa_f$ - effective thermal conductivity
!style-end!

!col-end!

!col! width=60%

!style! fontsize=80%
- $\alpha$ - volumetric heat-transfer coefficient (using correlations)
- $T_s$ - solid temperature
- $S_{ext}$ - volumetric source term
!style-end!

!col-end!

!row-end!

!---

> # The Navier-Stokes equations (contd.)

Reviewer 3: ```suggestion
# The Porous Navier-Stokes equations (contd.)
```

  style=font-size:26pt

The equations are extended with initial and boundary conditions:

- Few examples:

>   - Inlet: $\vec{u}=\vec{u}_{inlet}, \quad T=T_{inlet}, \quad \vec{r} \in \partial\Omega_{inlet}$

Reviewer 2: Velocity Inlet
Pressure Outlet

etc ?
since we can do different inlets (mass flux inlet, pressure inlet etc)

Author: Good idea!

  - Outlet: $p = p_{outlet}, \quad \vec{r} \in \partial\Omega_{outlet}$
  - No-slip, insulated walls: $\vec{u}=\vec{0}, \quad \kappa_f \nabla T \cdot \vec{n} = 0,  \quad \vec{r} \in \partial\Omega_{wall}$

>   - Heat flux on the walls: $\quad -\kappa_f \nabla T \big|_{\partial \Omega} \cdot \vec{n} = 0,  \quad \vec{r} \in \partial\Omega_{heated}$

Reviewer: ```suggestion
  - Heat flux on the walls: $\quad -\kappa_f \nabla T \big|_{\partial \Omega} \cdot \vec{n} = q\dprime,  \quad \vec{r} \in \partial\Omega_{heated}$
```

  - Free slip on walls: $\vec{u} \cdot \vec{n} = 0, \quad \vec{r} \in \partial\Omega_{slip}$
  - ...

!---

# Useful Acronyms

Good to know before navigating source code and documentation:

!row!
!col! width=50%

- +NS+ - +N+avier-+S+tokes
- +FV+ - +F+inite +V+olume
- +I+ - +I+ncompressible
- +WC+ - +W+eakly-+C+ompressible
- +C+ - +C+ompressible
- +P+ - +P+orous-Medium

!col-end!

!col! width=50%

Few Examples:

- +PINSFV:+ +P+orous-Medium +I+ncompressible +N+avier-+S+tokes +F+inite +V+olume
- +WCNSFV:+ +W+eakly-+C+ompressible +N+avier-+S+tokes +F+inite +V+olume

!col-end!

!row-end!

!---

# Building Input Files
style=font-size:28pt

The building blocks in MOOSE for terms in the PDEs are +Kernels+ for FE or +FVKernels+ for FV:

!style! fontsize=80%
!equation

> \underbrace{\frac{\partial \rho  \vec{u}}{\partial t}}_{\text{PINSFMomentumTimeDerivative}}

Reviewer 3: ```suggestion
\underbrace{\frac{\partial \rho  \vec{u}}{\partial t}}_{\text{PINSFVMomentumTimeDerivative}}
```

+ \underbrace{\nabla \cdot \left(\gamma^{-1}  \rho \vec{u} \otimes \vec{u}\right)}_{\text{PINSFVMomentumAdvection}} =

!equation
\underbrace{\nabla \cdot \left(\gamma \mu_\text{eff} \left(\nabla\vec{u}_I + \left(\nabla \vec{u}_I\right)^T-\frac{2}{3}\nabla\cdot\vec{u}_I\mathbb{I}\right)\right)}_{\text{PINSFVMomentumDiffusion}}
\underbrace{-\gamma\nabla p}_{\text{PINSFVMomentumPressure}}
+ \underbrace{\gamma\rho \vec{g}}_{\text{PINSFVMomentumGravity~}}
\underbrace{-  W  \vec{u}_I}_{\text{PINSFVMomentumFriction}}

The building blocks in MOOSE for boundary conditions are +BCs+ for FE or +FVBCs+ for FV:

- Inlet Velocity: $\text{INSFVInletVelocityBC}$
- Inlet Temperature: $\text{FVFunctionDirichletBC}$

> - Outlet pressure: $\text{INSFVOutletPRessureBC}$

Reviewer 3: ```suggestion
- Outlet pressure: $\text{INSFVOutletPressureBC}$
```

> - Heatflux: $\text{FVFunctionNeumannBC}$

Reviewer 2: this one isnt right

Author: Which one? I got this from the actin, do we need to change that too?

```
else if (_energy_wall_types[bc_ind] == "heatflux")
    {
      const std::string bc_type = "FVFunctionNeumannBC";
      InputParameters params = _factory.getValidParams(bc_type);
      params.set<NonlinearVariableName>("variable") = _fluid_temperature_name;
      params.set<FunctionName>("function") = _energy_wall_function[bc_ind];
      params.set<std::vector<BoundaryName>>("boundary") = {_wall_boundaries[bc_ind]};

      _problem->addFVBC(bc_type, _fluid_temperature_name + "_" + _wall_boundaries[bc_ind], params);
    }
```

- Freeslip: $\text{INSFVNaturalFreeSlipBC}$
- ...

!style-end!

!---

# A Simple Example: Laminar Free Flow in a Channel
style=font-size:26pt

!media navier_stokes/nsfv-channel-example.png style=width:100%;background:white;

Let us consider the following (fictional) material properties:

- $\mu=1.1$ $\text{Pa}\cdot\text{s}$

> - $\rho=1.1$ $\frac{kg}{m^3}$

Reviewer 2: ```suggestion
- $\rho=1.1$ $\frac{kg}{m^3}$ -> use incompressible formulation
```

!---

# Detailed Input File
style=font-size:26pt

!listing modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/2d-rc-no-slip.i

!---

> # The Navier-Stokes Finite Folume Action

Reviewer 3: ```suggestion
# The Navier-Stokes Finite Volume Action
```

style=font-size:26pt

> For the documentation of the action, click [here](NSFVAction.md)!

Reviewer 2: ```suggestion
A simplified syntax for the available has been developed, using the `action` system.
For the documentation of the action, click [here](NSFVAction.md)!
```

!listing modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/2d-rc-no-slip-action.i

!---

# Recommendations for Building Input Files
style=font-size:28pt

> 

Reviewer 2: use the action if possible

- Use Rhie-Chow interpolation for the advecting velocity

  - Other interpolation techniques may lead to checker-boarding/instability

- Start with first-order advected-interpolation schemes (e.g. upwind)
- Make sure that the pressure is pinned for incompressible/weakly-compressible simulations in close-loop systems
- For monolithic solvers (the default at the moment) use a variant of LU preconditioner
- For complex monolithic systems monitor the residuals of every variable

> 

Reviewer 2: keep elements below 1M ?? Or something like that 
@lindsayad what s the biggest you have ran

Reviewer: Increase `consistent_scaling` parameter in case of porous-media-driven checkerboard oscillations

Author: Got it, added!

Reviewer 3: I think I've run 10 million dofs. I don't know what the element count was

!---

> 

Reviewer 2: reference the current NS report? It s NEAMS TH something, in the citing page for MOOSE

Author: Can we cite the report from the end of last year? It had the Rhie-Chow interpolation, the porosity smoothing and all the other changes. Lemme check if I can find it.

Author: Or is it for internal use only?

Author: Added more references!

!style-end!

!---

# A Simple Example: Laminar Free Flow in a Channel
style=font-size:26pt

!media navier_stokes/nsfv-channel-example.png style=width:100%;background:white;

Let us consider the following (fictional) material properties:

- $\mu=1.1$ $\text{Pa}\cdot\text{s}$
- $\rho=1.1$ $\frac{kg}{m^3}$ (we are using incompressible formulation in this case)

!---

# Detailed Input File
style=font-size:26pt

!listing modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/2d-rc-no-slip.i

!---

# The Navier-Stokes Finite Volume Action
style=font-size:26pt

A simplified syntax has been developed, it relies on the `Action` system in MOOSE.
For the documentation of the corresponding action, click [here](NSFVAction.md)!

!listing modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/2d-rc-no-slip-action.i

!---

# Recommendations for Building Input Files
style=font-size:28pt

!style! fontsize=80%

- If possible, use the Navier Stokes Finite Volume Action syntax
- Use Rhie-Chow interpolation for the advecting velocity

  - Other interpolation techniques may lead to checker-boarding/instability

- Start with first-order advected-interpolation schemes (e.g. upwind)

> - Make sure that the pressure is pinned for incompressible/weakly-compressible simulations in close-loop systems

Reviewer 3: ```suggestion
- Make sure that the pressure is pinned for incompressible/weakly-compressible simulations in closed systems
```

> - For monolithic solvers (the default at the moment) use a variant of LU preconditioner

Reviewer 4: We need to explore more on this.

Author: This slideshow reflects our current knowledge. It is easy to append and extend as the module evolves. However, it would be important to merge this tonight because we need it for a training tomorrow :P.

Author: And thanks for reviewing it!

Reviewer 3: @fdkong did you look into splits at the end of last week?

- For complex monolithic systems monitor the residuals of every variable

> - Try to keep the number of elements relatively low (up to 1 million because LU preconditioner scales with $N^3$)

Reviewer 3: ```suggestion
- Try to keep the number of elements relatively low (up to 1 million because LU preconditioner scales with $N^3$). Larger problems can be solved in a reasonable wall-time as segregated-solve techniques are introduced
```

> - Try to keep the number of elements relatively low (up to 1 million because LU preconditioner scales with $N^3$). Larger problems can be solved in a reasonable wall-time as segregated-solve techniques are introduced

Reviewer 4: N3 might be for dense matrix. It is more complicated for sparse matrices. You might add a reference

Author: Okay, but the worst case scenario is N3 with spares too if the sparsity pattern is all over the place, not? I thought LU fills up the band between the diagonal and the entry furthest from it.

Reviewer 4: Right, the worse case would be 2/3 n^3. For sparse matrices, we might see a better performance than that in practice. However, direct solver is still  expensive even for "normal" sparse matrices.

Reviewer 4: How about this wordy sentences? You are free to say no

because the complexity of LU preconditioner is N^3 in the worse case,  its performance in general largely depends on the matrix sparsity pattern, and it is much more expensive compared with an iterative approach.

Author: I will add this as a footnote! Will push it in 5 minutes.

Author: Added!

Author: Thanks for the late-night work on this PR Fande!
